libs/vkd3d-shader/spirv.c

```diff
@@ -4486,6 +4486,41 @@ static unsigned int vkd3d_dxbc_compiler_get_output_variable_index(
     return register_idx;
 }
 
+static DWORD vkd3d_dxbc_compiler_double_source_mask_fixup(enum vkd3d_data_type src, enum vkd3d_data_type dst, DWORD mask)
+{
+    /* In DXBC, doubles take up a .xy whereas a normal float/int
+     * takes up a .x or a .y, etc.
+     * This maps a .xy writemask to a .x when going between
+     * double -> float/int and vice-versa for float/int -> double.
+     */
+
+    if (src == dst)
+        return mask;
+
+    if (src == VKD3D_DATA_DOUBLE)
+    {
+        unsigned int component_count = vkd3d_write_mask_component_count(mask);
+        if (component_count >= 2)
+            return VKD3DSP_WRITEMASK_0 | VKD3DSP_WRITEMASK_1 | VKD3DSP_WRITEMASK_2 | VKD3DSP_WRITEMASK_3;
+        else if (component_count == 1)
+            return VKD3DSP_WRITEMASK_0 | VKD3DSP_WRITEMASK_1;
+        else
+            return 0;
+    }
+    else if (dst == VKD3D_DATA_DOUBLE)
+    {
+        unsigned int component_count = vkd3d_write_mask_component_count(mask);
+        if (component_count >= 3)
+            return VKD3DSP_WRITEMASK_0 | VKD3DSP_WRITEMASK_1;
+        else if (component_count >= 1)
+            return VKD3DSP_WRITEMASK_0;
+        else
+            return 0;
+    }
+
+    return mask;
```

Comment:
The logic here seems wrong. Double instructions may operate on only the `.zw` components, in which case we would have to compress the mask to `VKD3DSP_WRITEMASK_1`.

Reply:
Double instructions can operate on .xy too.

The actual elements of the mask here don't matter anyway, it's ``popcount``'ed after so just the number of bits count.

This is just the mask for loading src regs.

Reply:
This isn't really clear from the code, and if we need a bit count rather than a mask, wouldn't it be better to return the bit count instead?

Reply:
The src reg loading thing takes in the writemask, similarly to DXVK.

Reply:
Consider it akin to https://github.com/doitsujin/dxvk/blob/master/src/dxbc/dxbc_compiler.cpp#L2947

Also I explained this in the comment in that function also.

Reply:
The comment explains what this does, but not why it is the correct thing to do, and I'm not convinced that it is. Take this for example:

```
    condition_mask = vkd3d_dxbc_compiler_double_source_mask_fixup(src[0].reg.data_type, dst->reg.data_type, dst->write_mask);
    condition_id = vkd3d_dxbc_compiler_emit_load_src(compiler, &src[0], condition_mask);
```
Now if the write mask is `.zw`, we load the *first* 32-bit component of the condition vector, which is wrong since we'd need to read the second one.

Reply:
> Now if the write mask is .zw, the resulting condition mask is VKD3DSP_WRITEMASK_0, which means that we load the first 32-bit component of the condition vector, which is wrong since we'd need to load the second one.

We'd need to load the first one regardless of whether the first element is selected or not.

``dadd r0.xyzw r1.xyzw r2.xyzw``
means ``r0.xyzw = r1.xyzw + r2.xyzw``

``dadd r0.zw r1.xyzw r2.xyzw``
means ``r0.zw = r1.xy + r2.xy``

Reply:
But this isn't about dadd, this is about dmovc.

Reply:
Hmmmmmmmmmmm, I will do more testing on this.

Reply:
fwiw it seems like a pain in the arse to convince fxc to generate fp64 code with .zw dst masks (although not necessarily impossible), but i'd think something like
```c
uint32_t out_mask = 0;
if (mask & VKD3DSP_WRITEMASK_0)
  out_mask |= VKD3DSP_WRITEMASK_0;
if (mask & VKD3DSP_WRITEMASK_2)
  out_mask |= VKD3DSP_WRITEMASK_1;
```
would be correct for the dmovc case.

I'm not so sure about the float conversions, the MSDN docs are weird for those. But `dtof` in particular is special because you can (in theory) have something like `dtof r0.w, r1.xyzw`, in which case I think we should be converting `r1.xy` anyway? At least I think that's what the logic behind the DXVK code was.

Reply:
At least for `ftod`, this logic is correct:
- In the scalar case, fxc only emits single-component swizzles anyway (i.e. `ftod r0.zw, r1.x`), which basically tells us that we'd need to look at the first component even if for some reason we end up getting a four-component swizzle.
- In the vec2 case, we get something like `ftod r0.xyzw, r1.yxyy`, so we only look at the first two components of the source operand.

Looking at [this](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dtof--sm5---asm-), it does sound like `dtof` basically does the reverse, so DXVK's (and by extension, this) logic is indeed correct for the float conversion case.

Reply:
So does this overall logic seem okay?

I think this satisfies all the citeria:

```c
static DWORD vkd3d_dxbc_compiler_double_source_mask_fixup(enum vkd3d_data_type src, enum vkd3d_data_type dst, DWORD mask)
{
    /* In DXBC, doubles take up a .xy whereas a normal float/int
     * takes up a .x or a .y, etc.
     * This maps a .xy writemask to a .x when going between
     * double -> float/int and vice-versa for float/int -> double.
     */

    if (src == dst)
        return mask;

    if (src == VKD3D_DATA_DOUBLE)
    {
        unsigned int component_count = vkd3d_write_mask_component_count(mask);
        if (component_count >= 2)
            return VKD3DSP_WRITEMASK_0 | VKD3DSP_WRITEMASK_1 | VKD3DSP_WRITEMASK_2 | VKD3DSP_WRITEMASK_3;
        else if (component_count == 1)
            return VKD3DSP_WRITEMASK_0 | VKD3DSP_WRITEMASK_1;
        else
            return 0;
    }
    else if (dst == VKD3D_DATA_DOUBLE)
    {
        DWORD out_mask = 0;
        if (mask & VKD3DSP_WRITEMASK_0)
            out_mask |= VKD3DSP_WRITEMASK_0;
        if (mask & VKD3DSP_WRITEMASK_2)
            out_mask |= VKD3DSP_WRITEMASK_1;
        return out_mask;
    }

    return mask;
}
```

Reply:
Committed the above for now.

Reply:
This breaks ftod and friends whereas the old logic was correct.

Can't you just use your original logic for those and treat cmovd as a special case? It doesn't make much sense to unify them in a single function since it behaves differently.

Reply:
Done.